### ISSUE TITLE
Devise.email_regexp is not correct

### ISSUE THREAD

**Issue by emn178** [heartcombo/devise#3820]
Hi,

It validates some incorrect emails as valid, like `emn178@gmail..com`

Change Devise.email_regexp
from

``` Ruby
/\A[^@\s]+@([^@\s]+\.)+[^@\W]+\z/
```

to

``` Ruby
/\A[\w+\-.]+@[a-z\d\-]+(\.[a-z]+)*\.[a-z]+\z/i
```

Refer to http://stackoverflow.com/questions/22993545/ruby-email-validation-with-regex

It can validate correctly.

**Comment by josevalim** [comment]
Yes, you can search previous issues on the topic but the regex is not meant to be a complete validation of the e-mail because it is much more complex than the regex you posted. For example, the `@` could be ip4 or ip6 addresses. So we decided to do a minimal validation. You can change it if you want.

**Comment by samjewell** [comment]
Ruby already has an email validation regex in `URI::MailTo::EMAIL_REGEXP`, and this is much stricter than the default in Devise, and also stricter than the various suggestions I've seen so far posted (on this thread, here and on #4268 and #5023)

```rb
URI::MailTo::EMAIL_REGEXP
=> /\A[a-zA-Z0-9.!\#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*\z/
```
From the [Ruby source](https://github.com/ruby/ruby/blob/ruby_2_6/lib/uri/mailto.rb#L55), this Regex is taken from the WHATWG's [_suggested_ regular expression](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address) for browsers to validate an email address as part of form validation.

So we patched this issue by adding:
```rb
  # Configure the email validator to be stricter
  config.email_regexp = URI::MailTo::EMAIL_REGEXP
```
to our `devise.rb` initializer.

@josevalim / @nashby / @ghost I feel this is a very rough edge on Devise, and one which every project/coder has to encounter and fix when their project grows large enough. Would you consider smoothing this rough edge, either by:
- Making the default validation itself stricter? (I saw somewhere a comment that this change is not backwards compatible, but could we make a major-version change for this)?
- Adding an additional comment to the `devise.rb` initialiser which is ready for other users to use out of the box, before they start persisting bad email addresses for the first time. Something like:
   ```rb
     # Configure the email validator to be stricter than the current and relatively permissive default
     # config.email_regexp = URI::MailTo::EMAIL_REGEXP
   ```
   
Let me know if you'll take either of these suggestions seriously - I can make a PR if that helps.

**Comment by carlosantoniodasilva** [comment]
Thanks for bringing this up again @samjewell.

I am hoping to work on a new major release at some point early this year, so that might be a good time to consider such a change in the Devise default. It'd also be possible to change this in the initialiser without causing any issue with existing apps; in other words, keeping the Devise default as is but changing how new apps are generated to the more strict version.

I'll reopen this for now to revisit when I have a chance to look into it in more detail / with more time.

**Comment by xtagon** [comment]
Before switching to `URI::MailTo::EMAIL_REGEXP` we saw quite a large number of invalid emails in production which resulted in unsuccesssful email delivery attempts. There's no way around it, users put in the wrong data sometimes. Sometimes a lot. We're talking about accidentally putting two dots instead of one in the domain name, accidentally typing = instead of ".", backticks, and somehow we had some with no-break-space (ASCII 160) characters which I think could be explained by users simply copying/pasting addresses.

`URI::MailTo::EMAIL_REGEXP` caught all of those cases.

This can all be validated on the client side as well, but when there's an API involved that's not enough, and IMO there is almost no benefit to keeping an incorrect e-mail validation just for simplicity's sake when it's no simpler than just using a valid regex. Yes it would warrant a major version bump, but is that such a bad thing? Semver exists so that breaking changes are explicit and known, not so that we can avoid making changes that would be breaking altogether.

Just my two cents (and huge thanks to everyone who provided solutions, it's appreciated!)

**Comment by nfedyashev** [comment]
I fully agree with @samjewell point. IMHO, tricter default email_regexp might make more sense for an average user of Devise.

In my case, "mailto:" prefix in user email was treated as a valid email and that was a bit surprising.

**Comment by samjewell** [comment]
> Are you guys handling this somehow as well? 

I've moved on from that job now, but IIRC we don't. Personally I'd be tempted not to bother. Maybe open an issue on the Ruby core, if that's what's not working properly.

**Comment by carlosantoniodasilva** [comment]
Thanks @samjewell. [Someone pointed out on Twitter](https://twitter.com/godfoca/status/1618066913510436864) that it's an actual valid use case for "email" as in `user@localhost` for example when using a local mail server. (I don't know about the RFC itself, and domain requirements, but from web app's perspective and devise, it's definitely not a valid one.)

**Comment by wTheRockb** [comment]
> Are you guys handling this somehow as well? 

I've changed my application to using `/[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]+/` (for forms when using `pattern` attribute) and `/\A[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]+/i` otherwise. We felt the benefit of disallowing `asdf@example` was stronger than allowing `user@localhost`.

**Comment by skreib** [comment]
hey, everyone!
are there any updates regarding this issue?

**Comment by fthobe** [comment]
@nfedyashev regarding “mailto:” validating

> In my case, "mailto:" prefix in user email was treated as a valid email and that was a bit surprising.

Inside a quoted string “:” is allowed, an example would be “john.doe”@example.com (yes, quoting the part before the @ is a valid legacy case. 

@carlosantoniodasilva 
> Thanks @samjewell. [Someone pointed out on Twitter](https://twitter.com/godfoca/status/1618066913510436864) that it's an actual valid use case for "email" as in user@localhost for example when using a local mail server. (I don't know about the RFC itself, and domain requirements, but from web app's perspective and devise, it's definitely not a valid one.)

Yes, this also used to be a valid legacy case. Anyhow most of these edge cases have been overturned by reality and are today mostly found in very old systems, a .local domain used to be another of these flavors. They come from a time in which local DNS management was at a very early stage. 

Today for all larger user systems such as Samba or Active Directory the usage of a valid domain is either encouraged or simply mandatory.
